### ISSUE TITLE
large memory usage

### ISSUE THREAD

**Issue by LzhinFdu** [zhuzilin/ring-flash-attention#23]
![image](https://github.com/zhuzilin/ring-flash-attention/assets/17453999/57e20774-f48c-47a6-8208-b97a23928b17)
Hello, thanks for sharing your awesome implementation of ring attention.
Here are my test results on 2*A100 (with nvlink).  Judging from the results, the memory usage of ring attention seems to be very large. This doesn't quite meet expectationss. Are there any possible problems?

**Comment by GeneZC** [comment]
The results of flash attention is somehow amazing... keep an eye on this.

**Comment by LzhinFdu** [comment]
> The results of flash attention are somehow amazing... keep an eye on this.

thanks, I'll check it

**Comment by GeneZC** [comment]
And after reading the code, I have found that the ring attention should accept already-chunked qkv instead of the whole qkv. That is, qkv should be split into local qkv before fed into the ring attention. This might be a diff.

**Comment by jzhang38** [comment]
@LzhinFdu @GeneZC Yeah you need to shard the sequence yourself before feeding them into ring-flash-attention. 
I have an implementation here: https://github.com/jzhang38/EasyContext

**Comment by LzhinFdu** [comment]
> And after reading the code, I have found that the ring attention should accept already-chunked qkv instead of the whole qkv. That is, qkv should be split into local qkv before fed into the ring attention. This might be a diff.
> 
> Not very certain that this is true, though.

That's right. Therefore, when comparing memory usage, the context used by Flash Attention should be doubled. Despite this, Flash Attention still maintains a significant lead.
